### ISSUE TITLE
ImapFolder events only raised when Folder Open?

### ISSUE THREAD

**Issue by DannyWave** [jstedfast/MailKit#718]
Hi There,

yesterday I replaced my very old static Mailkit Reference with the most recent nuget package.

What I first saw is that there is no longer a MessageArrived Event.
Furthermore after some testing it looks that events are only raised for the last opened folder!?

Before I could subscribe to all folders and got events for each of them.

Maybe something is completley wrong in the old fashioned way it worked before?

I know about Idle and NoOp, everything worked well before the upgrade.

**Comment by anthony-elhaddad** [comment]
hello, i just started to use mailkit and i am tying to find a way to keep my .net core web api listening to newly arrived messages. 
I went through the sample and found this Event: 

```
 client.Inbox.CountChanged += (sender, e) =>
    {
        // write logic here

     }
```

**Comment by jstedfast** [comment]
@DannyWave the MessagesArrived event was dropped because it had to do its own bookkeeping without all of the necessary information to be accurate. What you need to do is listen for the CountChanged event as well as the Expunged event, and based on that, figure out if new messages arrived when compared to your cached state.

As far as the events only being emitted on the most recently opened folder - that has always been the case. IMAP only allows 1 folder to be opened at a time. When you open a new folder, the previous folder is closed.

**Comment by jstedfast** [comment]
I'm pretty sure that Thunderbird does open multiple connections, but I doubt they open more than ~5 or so. I think I recall reading a blog post by them quite a few years ago now where they talked about opening up to 5 connections (I think they found that most servers didn't handle more than 4 or 5 connections well).

That said, you can always call `ImapFolder.Status (StatusItems.Unread);` to request that the ImapFolder's UnreadCount status gets updated and you don't need to open the folder to make that request.

If your server supports the `LIST-STATUS` extension (`ImapCapabilities.ListStatus`), then you can call `ImapFolder.GetSubfolders (StatusItems.Unread)` and it will be more efficient because it'll allow you to  update the status for a whole lost of folders at the same time. If the server doesn't support that extension, that API will still work, just know that MailKit will have to send a regular LIST command followed by a STATUS command for each folder returned which will likely be (slightly) slower than issuing a STATUS command for each folder yourself (assuming you already have the list of folders for earlier in the session).

**Comment by DannyWave** [comment]
Brilliant know how! Thanks again

**Comment by jstedfast** [comment]
In other words, instead of simply sending a `NOOP` command, you could call `STATUS` on each folder that is being displayed to the user.

Important note: `StatusItems` is a [Flags] enum, which means you can combine them to update multiple properties:

```csharp
folder.Status (StatusItems.Unread | StatusItems.Count);
```

**Comment by jstedfast** [comment]
No problem, glad to help :)

**Comment by DannyWave** [comment]
Hey Jeff, one more thing.

`Dim myMessages = pFolder.Fetch(0, -1, MessageSummaryItems.Full Or MessageSummaryItems.UniqueId Or MessageSummaryItems.BodyStructure)

            AddHandler pFolder.MessageExpunged, Sub(ByVal sender As Object, e As MailKit.MessageEventArgs)
                                                    If e.Index < myMessages.Count Then
                                                        Dim myMessage = myMessages(e.Index)
                                                        Debug.WriteLine("{0}: expunged message {1}: Subject: {2}", pFolder, e.Index, myMessage.Envelope.Subject)
                                                        'myMessages.RemoveAt(e.Index)
                                                        myMessages.Remove(myMessage)
                                                    Else
                                                        Debug.WriteLine("{0}: expunged message {1}: Unknown message.", pFolder, e.Index)
                                                    End If
                                                End Sub`

Both, myMessages.RemoveAt and myMessages.Remove result in client disconnect. Very strange. Did you come across that behaviour?

**Comment by jstedfast** [comment]
I've never seen that before. Are you getting some sort of exception there?

**Comment by DannyWave** [comment]
My bad. I put the using outer to have access on the done-Token as mentioned in your comments. Without try-catch there was no choice but disconnecting -.- Anyhow: After adding exceptionhandling there comes a message saying the collection is write protected!?

**Comment by DannyWave** [comment]
Ok, was missing `.ToList()` after the fetch... Problem solved

**Comment by DannyWave** [comment]
Hey Jeff,

I am having problems with re-fetching the folder when new messages arrived. 

The ImapClient is currently busy processing a command in another thread. Lock the SyncRoot property to properly synchronize your threads.

My code is as follows

```
Public Sub ActiveFolder(ByVal pFolder As ImapFolder)
    mDone = New CancellationTokenSource
    mActiveFolder = Nothing

    SyncLock pFolder.SyncRoot
        pFolder.Open(FolderAccess.ReadOnly)
        mMessages = pFolder.Fetch(0, -1, MessageSummaryItems.Full Or MessageSummaryItems.UniqueId Or MessageSummaryItems.BodyStructure).ToList
    End SyncLock
    mActiveFolder = pFolder

    Dim myThread As New Thread(AddressOf IdleState.IdleLoop)
    myThread.Start(New IdleState(mClient, mDone.Token, New CancellationToken))
    myThread.Join()
    Debug.WriteLine("End")
    mDone.Dispose()
End Sub

Private Sub mActiveFolder_CountChanged(sender As Object, e As EventArgs) Handles mActiveFolder.CountChanged
    Try
        Debug.WriteLine("The number of messages in {0} has changed.", mActiveFolder)
        If mActiveFolder.Count > mMessages.Count Then
            Debug.WriteLine("{0} new messages have arrived.", mActiveFolder.Count - mMessages.Count)

            mDone.Cancel()
            While mClient.IsIdle
                Thread.Sleep(100)
            End While

            SyncLock mClient.SyncRoot
                'Throws said exception
                mMessages = mActiveFolder.Fetch(0, -1, MessageSummaryItems.Full Or MessageSummaryItems.UniqueId Or MessageSummaryItems.BodyStructure).ToList
            End SyncLock
        End If
    Catch myEx As Exception
        Tracer.Trace(myEx)
    Finally
        ActiveFolder(mActiveFolder)
    End Try
End Sub
```


May you provide a furthermore example of how to exit idle, fetching the messages and enter idle again beatifully?

**Comment by jstedfast** [comment]
The problem is that you can't call back into the ImapFolder in an event callback because you cannot send a new command until the currently running command finishes execution (I should probably add this note to the Exception message because the SyncRoot info is not the only way this happens).

The way that MailKit's IMAP implementation is designed is that it will raise an event as it reads the server response rather than waiting until it reads the entire server response.

This means that you cannot send a new command in the event callback because there is more response data sent by the server that hasn't been read from the socket yet. If MailKit allowed you to send a new command, then it would get confused about the remaining response data from the previous command.

What you want to do is simply call `mDone.Cancel()` as you have done, and then return from the event handler.

Then, after `mThread.Join()`, call `Fetch()`.

**Comment by DannyWave** [comment]
> The way that MailKit's IMAP implementation is designed is that it will raise an event as it reads the server response rather than waiting until it reads the entire server response.

Makes sense for most cases. What about a beginReceive and endReceive event? Would help much in this cases

**Comment by jstedfast** [comment]
It's an unfortunate side-effect of MailKit's design, but that is correct - you can't do anything that will force the same ImapClient connection to query the server inside of an event callback.

I've thought about adding an event queue that the ImapClient would use and then flush the event queue only after it has finished reading/parsing the response but I haven't done that yet and I'm not sure if that would cause other problems (it might break Fetch() since users might connect to the MessageSummaryFetched event since those events need to be coordinated with the MessageExpunged and CountChanged events as well in order to keep your locale state fully in sync.

**Comment by DannyWave** [comment]
Therefore I changed it to

```
Private Sub mActiveFolder_CountChanged(sender As Object, e As EventArgs) Handles mActiveFolder.CountChanged
        Try
            Debug.WriteLine("The number of messages in {0} has changed.", mActiveFolder)
            If mActiveFolder.Count > mLocalMessages.Count Then
                Debug.WriteLine("{0} new messages have arrived.", mActiveFolder.Count - mLocalMessages.Count)

                mCallBack = New IdleCallBack(Sub()
                                                 'here comes the logic
                                             End Sub)

                Idle = False 'After Idle the mCallback is invoked
                Idle = True 'After mCallback invoked client will return in Idle State because of this flag
            End If
        Catch myEx As Exception
            Tracer.Trace(myEx)
        End Try
    End Sub
```

What do I need the said MessageSummaryFetched event for?

**Comment by jstedfast** [comment]
You likely don't need the MessageSummaryFetched event, but it is an event *other* people might use.

Sorry, it seems that my comment wasn't clear, I was talking in general terms about why MailKit works the way it does.

**Comment by DannyWave** [comment]
Actually I needed it because it is kind of async? Maybe what you said. Returning without having read to end!? I got some race conditions without waiting for that event.

After lots of questions it's time to say again how much I appreciate your work. I am very glad Mailkit exists! Thank you for all the support.
